### DISCUSSION TITLE
Can we have an option for gatsby-source-filesystem where it doesn't compute MD5 hashes since this can be slow with big files?

### DISCUSSION THREAD

**Discussion by FraserThompson** [gatsbyjs/gatsby#37425] [Ideas / Feature Requests] [answered]
I have a gatsby site with many large files (like almost a 1000 MP3s which are up to 120mb or so and 16,000 JPGs which are like 2-3mb before processing) and currently gatsby-source-filesystem chokes computing MD5 hashes for all these big files. This results in issues with Node open file limits (presumably because when each file takes a long time to hash, it means more will be opened simultaneously) and extremely long build times: 100 minutes currently, with 87 of those being in the `source and transform nodes` build step.

As far as I can tell the only purpose of the MD5 hashing is to determine if the file has changed. So I forked gatsby-source-filesystem and [made a variant](https://github.com/FraserThompson/gatsby-source-filesystem-fast) where instead of hashing the file for the `contentDigest` it just concatenates the file size in bytes and the modification time in milliseconds into a string. I figured this would be a pretty damn accurate "estimate" of whether the file has changed.

This brought my `source and transform nodes` time in a smaller test environment with only 3663 JPGs and 284 MP3s down from 781 to 10 seconds.

Does this seem like something which could be added into `gatsby-source-filesystem` , maybe as an option? I understand MD5 hashing is more robust, but in the vast majority of cases it seems like this approach wouldn't cause any issues and would significantly lower build times for people building sites with large files.

I'm not sure how many sites would benefit tbh since mine is pretty extreme, and I'm happy to maintain my fork for myself, but maybe it could be useful as an option for those edge cases.

**Comment by LekoArts** [comment]
Hi!

Thanks for the suggestion 👍 We currently use md5-file (https://github.com/kodie/md5-file/blob/master/index.js) which uses the standard Node.js `crypto` to create the hash. It might not be the fastest tho

Coming from https://stackoverflow.com/questions/1655769/fastest-md5-implementation-in-javascript?answertab=modifieddesc#tab-top, maybe give https://github.com/Daninet/hash-wasm at try in your fork (or one of the accepted code snippets - but unsure about the LICENSE there)?

Generally speaking I'd think an option would be fine (especially since you pointed out that the methodology is used in production elsewhere) but my preference would be to optimize this for everyone. If a faster md5 is not sufficient, then we can add the option :)

**Reply by FraserThompson** [reply]
I did actually try a different hashing library (murmurhash) which yielded some improvements (about 30%), but for my purposes it's still not fast enough and would still mean build times of over an hour.

hash-wasm looks like it performs even better: 494 seconds vs 781 seconds in my test environment. So a 45% improvement in this case.

Building the real site still runs into a too many open files error however, which I believe is caused by large files spending a long time being open while they're hashed (building without hashing doesn't run into this error). (I also forked https://github.com/FraserThompson/gatsby-source-filesystem-with-queue to solve this issue by limiting how many files it processes simultaneously).

hash-wasm sounds like an easy win though. It still means my full site build time is almost an hour but as I say my site is likely an extreme edge case.

I have changed my fork so hash-wasm is the default for free performance, and "no hashing" mode is an option with defined caveats (which is along the lines of what I'd expect to see if something similar were implemented in `gatsby-source-filesystem`).

**Reply by LekoArts** [reply]
> hash-wasm looks like it performs even better: 494 seconds vs 781 seconds in my test environment. So a 45% improvement in this case.

Nice 👍 

> Building the real site still runs into a too many open files error however, which I believe is caused by large files spending a long time being open while they're hashed (building without hashing doesn't run into this error). (I also forked [FraserThompson/gatsby-source-filesystem-with-queue](https://github.com/FraserThompson/gatsby-source-filesystem-with-queue) to solve this issue by limiting how many files it processes simultaneously).

I probably wouldn't want to introduce this to the main plugin if it can be solved with the "noHashing" option. It could be mentioned in the README.

Would you be willing to put up a PR for the official plugin so everyone can benefit?

We have instructions here and happy to help with the process along the way :) https://www.gatsbyjs.com/contributing/code-contributions/

**Reply by KyleAMathews** [reply]
Very interesting discussion! I agree size + mtime is good enough for most use cases. Perhaps even we make it the default for larger files since that's where the hashing slowdown happens?

Also looking at the benchmarks, SHA1 is a bit faster than MD5 it seems https://github.com/Daninet/hash-wasm#benchmark

**Reply by LekoArts** [reply]
If we can trust that size + mtime is always given we could also actually just change it out completely.

**Reply by KyleAMathews** [reply]
I think we could take some inspiration from Zig's caching system https://ziglang.org/download/0.4.0/release-notes.html#Build-Artifact-Caching — they use inode + mtime. They also check what the filesystem's mtime granularity is as some fs' are higher. 

mtime isn't 100% reliable as this blog post goes deep into https://apenwarr.ca/log/20181113 — but Zig's approach of storing the file hash in the cache & only regenerating it if mtime/inode change (or mtime looks suspicious) seems like a safe general approach that'd give everyone a speedup & then still have the option of skipping hashing (which is safe on a modern OS which is most people) for people that need it like @FraserThompson

**Reply by KyleAMathews** [reply]
so quick code:
```ts
const key = statObj.mtime + statObj.inode
let hash = await cache.get(key)
if (!hash) {
   hash = generateHash()
   await cache.set(key, hash)
}

const contentDigest = key + hash + pathToFile
```

**Reply by LekoArts** [reply]
@FraserThompson Would you be willing to work on this or should I take this? Don't want to step on your toes if you want to contribute :)

**Reply by FraserThompson** [reply]
@LekoArts Sure, I'd be happy to help. I'll start working on a pull request!

**Reply by FraserThompson** [reply]
Actually, here's an alternative idea for the "noHash" (working title) mode... I just noticed what gatsby-source-filesystem is doing for the contentDigest of directories:

```
    const contentDigest = createContentDigest({
      stats: stats,
      absolutePath: slashedFile.absolutePath,
    })
```

It passes an object containing the entire output of fs.stat and the absolute path to createContentDigest, which uses [node-object-hash](https://www.npmjs.com/package/node-object-hash) to return an MD5 hash of that object. This same approach could be probably be used for files?

I tested it and it's a tad slower than just using inode + mtime  (14 seconds vs 10 seconds), but still a lot faster than hashing each entire file (781 seconds), so it still solves the problem of large files causing massive build times.

Thought it was worth bringing up in case we'd prefer to use an already implemented and well-tested method for making contentDigest's instead of adding another one (even if it isn't the absolute fastest)?

**Reply by LekoArts** [reply]
Thanks for bringing that up! Let's first use inode + mtime and if that causes problems in some way we could try the object hashing :) I'll also think that my PR https://github.com/gatsbyjs/gatsby/pull/37433 gets merged today so you can then use hash-wasm's hashing function in your PR

And let's not use a negated option, we made that mistake in the past. We'd prefer "enable" options, so working title could be "fastHashing"

**Comment by LekoArts** [comment, marked as answer]
The PR https://github.com/gatsbyjs/gatsby/pull/37464 will be available in Gatsby 5.5 🎉
